modules/dnn/src/layers/einsum_layer.cpp

```diff
@@ -0,0 +1,1200 @@
+    inputs_arr.getMatVector(inputs);
+
+
+    preProcessedInputs.reserve(inputs.size());
+    homogenizedInputDims.reserve(inputs.size());
+
+    int inputIter = 0;
+    for(const Mat& input : inputs)
+    {
+        Mat preprocessed;
+
+        // variable to hold processed version of the original input
+        MatShape input_dims = shape(input);
+
+        const auto& currSubscriptIndices = inputSubscriptIndices[inputIter];
+
+        // There should be subscript index (subscript label) for each dim of the input
+        if (input_dims.size() != currSubscriptIndices.size())
+        {
+            CV_Error(Error::StsError,
+                "Rank of the input must match number of subscript labels corresponding to the input");
+        }
+
+        std::vector<int> subscriptIndicesToInputIndex(numLetterIndices, -1);
+        // this will hold input dims after reordering so that all inputs have
+        // same axes order
+        MatShape homogenizedInputDims_(numLetterIndices, 1);
+
+        int dimIndexInIreprocessedInput = 0;
+        int dimIndexInOriginalInput = 0;
+
+        for (const auto& subscriptIndex : currSubscriptIndices)
+        {
+            if(subscriptIndicesToInputIndex[subscriptIndex] == -1)
+            {
+                subscriptIndicesToInputIndex[subscriptIndex] = dimIndexInIreprocessedInput++;
+                homogenizedInputDims_[subscriptIndex] = input_dims[dimIndexInOriginalInput];
+            } else {
+                // Call diagonal
+                preprocessed = Diagonal(
+                    !preprocessed.empty() ? preprocessed : inputs[inputIter],
+                    subscriptIndicesToInputIndex[subscriptIndex],
+                    dimIndexInIreprocessedInput);
+            }
+            ++dimIndexInOriginalInput;
+        }
+
+        std::vector<size_t> permutation;
+        for(auto& d : subscriptIndicesToInputIndex)
+        {
+            if (d != -1)
+                permutation.emplace_back(d);
+        }
+
+        if (IsTransposeRequired(
+            !preprocessed.empty() ? preprocessed.size.dims() : inputs[inputIter].size.dims(),
+            permutation))
+        {
+            // call transpose
+            preprocessed = Transpose(
+                !preprocessed.empty() ? preprocessed : inputs[inputIter],
+                !preprocessed.empty() ? shape(preprocessed) : shape(inputs[inputIter]),
+                permutation);
+        }
+
+        if (!preprocessed.empty())
+        {
+            // check if this is correct
+            preprocessed = preprocessed.reshape(1, homogenizedInputDims_.size(), homogenizedInputDims_.data());
+        }
+
+        // fails here! check the problem
+        preProcessedInputs.emplace_back(preprocessed);
+        homogenizedInputDims.emplace_back(homogenizedInputDims_);
+
+        ++inputIter;
+    }
+}
+
+void LayerEinsumImpl::parseEquation(const String& equation)
+{
+    // copy copy of an eqution, will be changed
+    String eq = equation;
```

Comment:
In case, if you need a copy, you can replace `const String& equation` with  `String equation` in parameters. Copy is done automatically by compiler.